mysql/azuremysql/azuremysql.go

```diff
@@ -0,0 +1,127 @@
+	"fmt"
+	
+	"contrib.go.opencensus.io/integrations/ocsql"
+	"github.com/go-sql-driver/mysql"	
+)
+
+const (
+	defaultPort = 3306
+	endpointSuffix = "mysql.database.azure.com"
+)
+
+// Params specifies how to connect to an Azure Database for MySQL.
+type Params struct {
+	// ServerName is the MYSQL instance name without domain suffix. Example: gocloud
+	ServerName string	
+	// User is the database user to connect as.
+	User string
+	// Password is the database user password to use.
+	Password string
+	// Database is the MYSQL database name to connect to.
+	Database string
+	// TraceOpts contains options for OpenCensus.
+	TraceOpts []ocsql.TraceOption
+}
+// GetFQDN constructs the FQDN for Azure Database for MySQL.
+func (p *Params) GetFQDN() string {
+	fqdn := fmt.Sprintf("%s.%s:%v", p.ServerName, endpointSuffix, defaultPort) 
+	return fqdn;
+}
+// Validate ensures all required parameters are set.
+func (p *Params) Validate() error {
+	if p.ServerName == "" || p.User == "" || p.Database == "" {
+		return fmt.Errorf("Missing one or more required params; got servername=%q username=%q database=%q", p.ServerName, p.User, p.Database)
+	}
+	return nil
+}
+// Open opens an encrypted connection to an Azure Database for MySql database.
+func Open(ctx context.Context, acf *AzureCertFetcher, params *Params) (*sql.DB, func(), error) {	
+	if e := params.Validate(); e != nil {
+		return nil, nil, e
+	}
+	c := &connector{		
+		provider: *acf,
+		params:   *params,
+		sem:      make(chan struct{}, 1),
+		ready:    make(chan struct{}),
+	}
+	c.sem <- struct{}{}
+	// Make a copy of TraceOpts to avoid caller modifying.
+	c.params.TraceOpts = append([]ocsql.TraceOption(nil), c.params.TraceOpts...)
+
+	db := sql.OpenDB(c)
+	return db, func() { db.Close() }, nil
+}
+type connector struct {
+	params Params
+
+	sem      chan struct{} // receive to acquire, send to release	
+	provider AzureCertFetcher
+
+	ready chan struct{} // closed after writing dsn
+	dsn   string
+}
+
+func (c *connector) Connect(ctx context.Context) (driver.Conn, error) {
+	select {
+	case <-c.sem:		
+		certPool, err := c.provider.AzureCertPool(ctx)
+		if err != nil {
+			c.sem <- struct{}{} // release
+			return nil, fmt.Errorf("connect Azure MySql: %v", err)
+		}
+		
+		tlsConfigName := fmt.Sprintf("gocloud.dev/mysql/azuresql/%s", c.params.ServerName)
+		err = mysql.RegisterTLSConfig(tlsConfigName, &tls.Config{
+			RootCAs: certPool,
+		})
+		if err != nil {
+			c.sem <- struct{}{} // release
+			return nil, fmt.Errorf("connect Azure MySql: register TLS: %v", err)
+		}
+		cfg := &mysql.Config{
+			Net:                     "tcp",
+			Addr:                    c.params.GetFQDN(),
+			User:                    c.params.User,
+			Passwd:                  c.params.Password,
+			TLSConfig:               tlsConfigName,
+			AllowCleartextPasswords: true,
+			AllowNativePasswords:    true,
+			DBName:                  c.params.Database,
+		}
+		c.dsn = cfg.FormatDSN()
+		close(c.ready)
+
+	case <-c.ready:
```

Comment:
Can you update with a couple of explanatory comments, similar to https://github.com/google/go-cloud/blob/master/mysql/rdsmysql/rdsmysql.go#L121 and https://github.com/google/go-cloud/blob/master/mysql/rdsmysql/rdsmysql.go#L123, for clarity?

Reply:
Ok, comments added.